src/main/scala/rise/core/types/Solution.scala

```diff
@@ -104,6 +104,7 @@ case class Solution(ts: Map[Type, Type],
       case lambda@NatToNatLambda(x, body) =>
         val xSub = apply(x) match {
           case n: NatIdentifier => n
+          case n: arithexpr.arithmetic.NamedVar => NatIdentifier(n.name, n.range, isExplicit = true)
```

Comment:
when does this happen? that might indicate a bug elsewhere, since the `NatIdentifier` information should not be lost.

Reply:
This addition was definitely required to get the code to work. I would suggest that this is an orthogonal issue top this PR, as I basically copied this solution from the handling of `NatToData` that matches for `NamedVar` as well.

Reply:
Maybe you can create an issue for this and we merge this PR?